src/lightning/pytorch/loops/fit_loop.py

```diff
@@ -244,12 +244,28 @@ def setup_data(self) -> None:
         if trainer.datamodule is not None:
             allow_zero_length |= trainer.datamodule.allow_zero_length_dataloader_with_multiple_devices
 
+        limits = []
+        for dl in combined_loader.flattened:
+            # determine number of batches
+            length = len(dl) if has_len_all_ranks(dl, trainer.strategy, allow_zero_length) else float("inf")
+            num_batches = _parse_num_batches(stage, length, trainer.limit_train_batches)
+            limits.append(num_batches)
+
+        combined_loader.limits = limits
+
+        training = trainer.training
+        trainer.training = True
+        self._data_fetcher = _select_data_fetcher(trainer)
+        trainer.training = training
+
+        self._data_fetcher.setup(combined_loader)
+        iter(self._data_fetcher)  # creates the iterator inside the fetcher
+        max_batches = sized_len(combined_loader)
+        self.max_batches = max_batches if max_batches is not None else float("inf")
         has_len_all_ranks_ = has_len_all_ranks(combined_loader, trainer.strategy, allow_zero_length)
-        self.max_batches = len(combined_loader) if has_len_all_ranks_ else float("inf")
-        if self.max_batches == 0:
-            return
 
-        self.max_batches = _parse_num_batches(stage, self.max_batches, trainer.limit_train_batches)
+        if sum(limits) == 0:
```

Comment:
Since you have self.max_batches properly defined again, this could be reverted to what was here before: `if self.max_batches == 0`